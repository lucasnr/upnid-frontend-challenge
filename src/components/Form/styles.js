import styled from 'styled-components';

export const Container = styled.form`
	align-items: center;
	display: flex;
	margin-bottom: 1.5rem;
	width: 20rem;

	input {
		border: none;
		border-radius: 0.125rem;
		font-size: 0.75rem;
		flex: 1;
		margin-right: 0.75rem;
		width: 100%;
		padding: 0.375rem 1rem;
	}

	button {
		background-color: #49e;
		border: none;
		border-radius: 0.125rem;
		color: #fff;
		font-size: 0.75rem;
		padding: 0.375rem 1rem;
		text-transform: uppercase;

		span {
			display: none;
		}
	}

	@media (min-width: 576px) {
		width: 30rem;

		input,
		button {
			font-size: 1rem;

			span {
				display: inline;
			}
		}
	}
`;
