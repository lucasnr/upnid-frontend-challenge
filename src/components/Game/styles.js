import styled, { keyframes } from 'styled-components';

import bg from '../../assets/img/bg.gif';
import bgPaused from '../../assets/img/bg-pause.png';
import car from '../../assets/img/car.png';
import obstacle from '../../assets/img/obstacle.png';

export const Container = styled.main`
	background-image: url(${(props) => (props.paused ? bgPaused : bg)});
	background-position: center;
	background-size: cover;
	border-radius: 0.25rem;
	height: 20rem;
	overflow: hidden;
	position: relative;
	width: 20rem;

	@media (min-width: 576px) {
		height: 30rem;
		width: 30rem;
	}
`;

export const Car = styled.img.attrs((props) => ({
	alt: 'Car icon',
	style: {
		left: `${props.left}%`,
	},
	src: car,
}))`
	bottom: 0px;
	height: 6rem;
	position: absolute;
	transition: left 300ms ease;
	width: 6rem;

	@media (min-width: 576px) {
		height: 9rem;
		width: 9rem;
	}
`;

const obstacleAnimation = (props) => keyframes`
	0% {
		height: .625rem;
		top: 45%;
		width: 1rem;
	}

	100% {
		height: 7.25rem;
		left: ${props.to}%;
		top: 100%;
		width: 8rem;
	}
`;

export const Obstacle = styled.img.attrs((props) => ({
	alt: 'Obstacle icon',
	style: {
		left: `${props.initial || 45}%`,
	},
	src: obstacle,
}))`
	animation: ${(props) => obstacleAnimation(props)} 2s linear forwards;
	position: absolute;
	transition: left 300ms ease;
`;

export const Paused = styled.span.attrs((props) => ({
	children: 'Paused',
}))`
	background-color: rgba(0, 0, 0, 0.75);
	border-radius: 0.25rem;
	color: #fff;
	font-size: 1.5rem;
	font-weight: bold;
	left: 50%;
	padding: 1rem 2rem;
	position: absolute;
	top: 50%;
	text-transform: uppercase;
	transform: translate(-50%, -50%);
`;
