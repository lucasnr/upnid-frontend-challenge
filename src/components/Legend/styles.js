import styled from 'styled-components';

export const Container = styled.section`
	margin-top: 1rem;
	width: 20rem;

	@media (min-width: 576px) {
		width: 30rem;
	}
`;

export const Name = styled.h2`
	font-size: 1rem;
	font-weight: normal;
	text-align: center;
`;

export const Controls = styled.div`
	display: grid;
	grid-template-columns: repeat(2, 1fr);
	margin-top: 1rem;

	& > * {
		font-size: 0.75rem;

		&:nth-child(even) {
			text-align: right;
		}
	}
`;
