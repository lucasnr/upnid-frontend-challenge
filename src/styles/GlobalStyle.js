import { createGlobalStyle } from 'styled-components';

const GlobalStyle = createGlobalStyle`
	* {
		box-sizing: border-box;
		margin: 0px;
		outline: none;
		padding: 0px;
	}

	ul {
		list-style: none;
	}

	a {
		color: inherit;
		text-decoration: none;
	}

	body {
		background-color: #efefef;
	}

	body, input, button {
		font-family: 'Poppins', sans-serif;
	}

	button {
		cursor: pointer;
	}

	#root {
		align-items: center;
		display: flex;
		justify-content: center;
		min-height: 100vh;
	}
`;

export default GlobalStyle;
