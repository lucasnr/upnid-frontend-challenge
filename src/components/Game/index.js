import React, { useEffect, useState } from 'react';

import { Container, Car } from './styles';

export default function Game() {
	const [left, setLeft] = useState(33.33);
	useEffect(() => {
		document.onkeydown = function (e) {
			e = e || window.event;
			if (e.key === 'ArrowRight')
				setLeft((prev) => (prev === 66.66 ? prev : prev + 33.33));
			else if (e.key === 'ArrowLeft')
				setLeft((prev) => (prev === 0 ? prev : prev - 33.33));
			else if (e.key.toLowerCase() === 'd') setLeft(66.66);
			else if (e.key.toLowerCase() === 'a') setLeft(0);
			else if (e.key.toLowerCase() === 's') setLeft(33.33);
		};

		let xDown = null;
		let yDown = null;

		document.addEventListener('touchstart', handleTouchStart, false);
		document.addEventListener('touchmove', handleTouchMove, false);

		const getTouches = (evt) => evt.touches || evt.originalEvent.touches;
		function handleTouchStart(evt) {
			const firstTouch = getTouches(evt)[0];
			xDown = firstTouch.clientX;
			yDown = firstTouch.clientY;
		}

		function handleTouchMove(evt) {
			if (!xDown || !yDown) return;

			let xUp = evt.touches[0].clientX;
			let yUp = evt.touches[0].clientY;

			let xDiff = xDown - xUp;
			let yDiff = yDown - yUp;

			if (Math.abs(xDiff) > Math.abs(yDiff)) {
				if (xDiff > 0) setLeft((prev) => (prev === 0 ? prev : prev - 33.33));
				else setLeft((prev) => (prev === 66.66 ? prev : prev + 33.33));
			}

			xDown = null;
			yDown = null;
		}
	}, []);

	return (
		<Container>
			<Car left={left} />
		</Container>
	);
}
