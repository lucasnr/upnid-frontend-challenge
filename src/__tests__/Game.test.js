import React from 'react';
import { shallow } from 'enzyme';

import Game from '../components/Game';
import { Message } from '../components/Game/styles';

describe('Testing game component', () => {
	it('renders without crashing and matches snapshot', () => {
		const wrapper = shallow(<Game />);
		expect(wrapper).toMatchSnapshot();

		wrapper.setProps({ start: true });
		expect(wrapper).toMatchSnapshot();

		expect(wrapper.find(Message).text()).toBe('3');
	});
});
