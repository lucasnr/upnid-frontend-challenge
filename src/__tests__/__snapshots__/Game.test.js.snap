// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing game component renders without crashing and matches snapshot 1`] = `
<styled.main
  paused={true}
>
  <styled.img
    left={33.33}
  />
</styled.main>
`;

exports[`Testing game component renders without crashing and matches snapshot 2`] = `
<styled.main
  paused={false}
>
  <styled.img
    left={33.33}
  />
  <styled.span
    countdown={3}
  >
    3
  </styled.span>
</styled.main>
`;
